vnext/Microsoft.ReactNative/Views/TouchEventHandler.cpp

```diff
@@ -506,26 +506,27 @@ bool TouchEventHandler::TagFromOriginalSource(
 
   // Find the React element that triggered the input event
   xaml::UIElement sourceElement = args.OriginalSource().try_as<xaml::UIElement>();
-  winrt::IPropertyValue tag(nullptr);
+  int64_t tag = -1;
 
   while (sourceElement) {
-    auto tagValue = sourceElement.ReadLocalValue(xaml::FrameworkElement::TagProperty());
-    if (tagValue != xaml::DependencyProperty::UnsetValue()) {
-      tag = tagValue.try_as<winrt::IPropertyValue>();
+    tag = GetTag(sourceElement);
+    if (tag != -1) {
       // If a TextBlock was the UIElement event source, perform a more accurate hit test,
       // searching for the tag of the nested Run/Span XAML elements that the user actually clicked.
       // This is to support nested <Text> elements in React.
       // Nested React <Text> elements get translated into nested XAML <Span> elements,
       // while the content of the <Text> becomes a list of XAML <Run> elements.
       // However, we should report the Text element as the target, not the contexts of the text.
       if (const auto textBlock = sourceElement.try_as<xaml::Controls::TextBlock>()) {
-        const auto pointerPos = args.GetCurrentPoint(textBlock).RawPosition();
-        const auto inlines = textBlock.Inlines().GetView();
+        if (textBlock.Inlines().Size() == 0) {
```

Comment:
this whole dance of having the touch handler know about UI elements seems backwards which is why it seems better to have the touch event handler query the view manager to see if it wants to do any special handling, and encapsulating that logic inside the VM

Reply:
Totally agree. This is not fundamentally different from the "current" implementation of text hit testing though, so will separate concerns and leave as future TODO to create something analogous to the `IReactCompoundView` we had in C# RNW, (or the `IViewManagerWithHitTesting` you proposed above).